File: android/src/main/java/com/margelo/nitro/stroketext/HybridStrokeTextView.kt
package com.margelo.nitro.stroketext

import android.view.View
import com.facebook.jni.HybridData
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.RecyclableView

class HybridStrokeTextView(context: ThemedReactContext) : HybridStrokeTextViewSpec(), RecyclableView {
  private val strokeTextView = StrokeTextView(context)
  override val view: View = strokeTextView

  private var isDisposed = false

  override var text: String = ""
  override var color: String? = null
  override var strokeColor: String? = null
  override var strokeWidth: Double? = null
  override var fontSize: Double? = null
  override var fontWeight: String? = null
  override var fontFamily: String? = null
  override var fontStyle: StrokeTextFontStyle? = null
  override var lineHeight: Double? = null
  override var letterSpacing: Double? = null
  override var textAlign: StrokeTextAlign? = null
  override var textDecorationLine: StrokeTextDecorationLine? = null
  override var textTransform: StrokeTextTransform? = null
  override var opacity: Double? = null
  override var allowFontScaling: Boolean? = null
  override var maxFontSizeMultiplier: Double? = null
  override var includeFontPadding: Boolean? = null
  override var numberOfLines: Double? = null
  override var ellipsizeMode: StrokeTextEllipsizeMode? = null
  override var padding: Double? = null
  override var paddingVertical: Double? = null
  override var paddingHorizontal: Double? = null
  override var paddingTop: Double? = null
  override var paddingRight: Double? = null
  override var paddingBottom: Double? = null
  override var paddingLeft: Double? = null

  override fun dispose() {
    if (isDisposed) return
    isDisposed = true
    super.dispose()

    try {
      val hybridDataField = HybridStrokeTextViewSpec::class.java.getDeclaredField("mHybridData")
      hybridDataField.isAccessible = true
      val hybridData = hybridDataField.get(this) as? HybridData ?: return

      val resetNativeMethod = HybridData::class.java.getDeclaredMethod("resetNative")
      resetNativeMethod.isAccessible = true
      resetNativeMethod.invoke(hybridData)
    } catch (_: Throwable) {
    }
  }

  override fun prepareForRecycle() {
    text = ""
    color = null
    strokeColor = null
    strokeWidth = null
    fontSize = null
    fontWeight = null
    fontFamily = null
    fontStyle = null
    lineHeight = null
    letterSpacing = null
    textAlign = null
    textDecorationLine = null
    textTransform = null
    opacity = null
    allowFontScaling = null
    maxFontSizeMultiplier = null
    includeFontPadding = null
    numberOfLines = null
    ellipsizeMode = null
    padding = null
    paddingVertical = null
    paddingHorizontal = null
    paddingTop = null
    paddingRight = null
    paddingBottom = null
    paddingLeft = null

    afterUpdate()
  }

  override fun afterUpdate() {
    val displayMetrics = strokeTextView.resources.displayMetrics
    val resolvedAllowFontScaling = allowFontScaling ?: true
    val resolvedMaxFontSizeMultiplier =
        maxFontSizeMultiplier
            ?.toFloat()
            ?.takeIf { it.isFinite() && it >= 1f }

    strokeTextView.rawText = text

    strokeTextView.color =
        StrokeTextView.parseColor(color) ?: strokeTextView.resolvedDefaultTextColor()
    strokeTextView.strokeColor = StrokeTextView.parseColor(strokeColor) ?: android.graphics.Color.TRANSPARENT
    val resolvedStrokeWidth = (strokeWidth ?: 0.0).coerceAtLeast(0.0)
    strokeTextView.strokeWidthDp = resolvedStrokeWidth
    strokeTextView.strokeWidthPx =
        StrokeTextView.dpToPx(
            resolvedStrokeWidth.toFloat(),
            displayMetrics,
        )

    strokeTextView.fontSizePx =
        StrokeTextView.textToPx(
            fontSize ?: 14.0,
            resolvedAllowFontScaling,
            resolvedMaxFontSizeMultiplier,
            displayMetrics,
        )
    strokeTextView.fontWeight = fontWeight
    strokeTextView.fontFamily = fontFamily
    strokeTextView.fontStyle = fontStyle ?: StrokeTextFontStyle.NORMAL
    strokeTextView.lineHeightPx =
        lineHeight?.let {
          StrokeTextView.textToPx(
              it,
              resolvedAllowFontScaling,
              resolvedMaxFontSizeMultiplier,
              displayMetrics,
          )
        }
    strokeTextView.letterSpacingPx =
        letterSpacing?.let {
          StrokeTextView.textToPx(
              it,
              resolvedAllowFontScaling,
              resolvedMaxFontSizeMultiplier,
              displayMetrics,
          )
        }

    strokeTextView.textAlign = textAlign ?: StrokeTextAlign.AUTO
    strokeTextView.textDecorationLine = textDecorationLine ?: StrokeTextDecorationLine.NONE
    strokeTextView.textTransform = textTransform ?: StrokeTextTransform.NONE

    strokeTextView.includeFontPadding = includeFontPadding ?: false
    strokeTextView.numberOfLines = (numberOfLines ?: 0.0).toInt()
    strokeTextView.ellipsizeMode = ellipsizeMode

    strokeTextView.paddingAllPx = padding?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingVerticalPx = paddingVertical?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingHorizontalPx = paddingHorizontal?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingTopPx = paddingTop?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingRightPx = paddingRight?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingBottomPx = paddingBottom?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }
    strokeTextView.paddingLeftPx = paddingLeft?.let { StrokeTextView.dpToPx(it.toFloat(), displayMetrics) }

    strokeTextView.alpha = (opacity ?: 1.0).toFloat()

    strokeTextView.invalidateTextLayout()
  }
}

File: ios/HybridStrokeTextView.swift
import NitroModules
import UIKit

final class HybridStrokeTextView: HybridStrokeTextViewSpec {
  typealias ViewType = StrokeTextView

  let view = StrokeTextView()

  var text: String = ""
  var color: String? = nil
  var strokeColor: String? = nil
  var strokeWidth: Double? = nil
  var fontSize: Double? = nil
  var fontWeight: String? = nil
  var fontFamily: String? = nil
  var fontStyle: StrokeTextFontStyle? = nil
  var lineHeight: Double? = nil
  var letterSpacing: Double? = nil
  var textAlign: StrokeTextAlign? = nil
  var textDecorationLine: StrokeTextDecorationLine? = nil
  var textTransform: StrokeTextTransform? = nil
  var opacity: Double? = nil
  var allowFontScaling: Bool? = nil
  var maxFontSizeMultiplier: Double? = nil
  var includeFontPadding: Bool? = nil
  var numberOfLines: Double? = nil
  var ellipsis: Bool? = nil
  var padding: Double? = nil
  var paddingVertical: Double? = nil
  var paddingHorizontal: Double? = nil
  var paddingTop: Double? = nil
  var paddingRight: Double? = nil
  var paddingBottom: Double? = nil
  var paddingLeft: Double? = nil

  func afterUpdate() {
    view.text = text

    if let color = color, let parsed = StrokeTextColor.parse(color) {
      view.color = parsed
    } else {
      view.color = StrokeTextView.defaultTextColor()
    }

    if let strokeColor = strokeColor, let parsed = StrokeTextColor.parse(strokeColor) {
      view.strokeColor = parsed
    } else {
      view.strokeColor = .clear
    }

    let resolvedStrokeWidth = max(0, strokeWidth ?? 0)
    view.strokeWidth = CGFloat(resolvedStrokeWidth)

    view.fontSize = CGFloat(fontSize ?? 14)
    view.fontWeight = fontWeight ?? "400"
    view.fontFamily = fontFamily
    view.fontStyle = fontStyle ?? .normal

    view.allowFontScaling = allowFontScaling ?? true
    if let multiplier = maxFontSizeMultiplier, multiplier.isFinite, multiplier >= 1 {
      view.maxFontSizeMultiplier = CGFloat(multiplier)
    } else {
      view.maxFontSizeMultiplier = nil
    }

    view.lineHeight = lineHeight.map { CGFloat($0) }
    view.letterSpacing = letterSpacing.map { CGFloat($0) }
    view.textAlign = textAlign ?? .auto
    view.textDecorationLine = textDecorationLine ?? .none
    view.textTransform = textTransform ?? .none

    view.numberOfLines = Int(numberOfLines ?? 0)
    view.ellipsis = ellipsis ?? false

    view.alpha = CGFloat(opacity ?? 1)

    let baseInsets = resolvedPaddingInsets(
      padding: padding,
      paddingVertical: paddingVertical,
      paddingHorizontal: paddingHorizontal,
      paddingTop: paddingTop,
      paddingRight: paddingRight,
      paddingBottom: paddingBottom,
      paddingLeft: paddingLeft
    )
    let strokeInset = CGFloat(ceil(resolvedStrokeWidth) / 2.0)
    view.paddingInsets = UIEdgeInsets(
      top: baseInsets.top + strokeInset,
      left: baseInsets.left + strokeInset,
      bottom: baseInsets.bottom + strokeInset,
      right: baseInsets.right + strokeInset
    )
  }

  private func resolvedPaddingInsets(
    padding: Double?,
    paddingVertical: Double?,
    paddingHorizontal: Double?,
    paddingTop: Double?,
    paddingRight: Double?,
    paddingBottom: Double?,
    paddingLeft: Double?
  ) -> UIEdgeInsets {
    func resolve(_ specific: Double?, _ axis: Double?, _ all: Double?) -> CGFloat {
      return CGFloat(specific ?? axis ?? all ?? 0)
    }

    let top = resolve(paddingTop, paddingVertical, padding)
    let bottom = resolve(paddingBottom, paddingVertical, padding)
    let left = resolve(paddingLeft, paddingHorizontal, padding)
    let right = resolve(paddingRight, paddingHorizontal, padding)
    return UIEdgeInsets(top: top, left: left, bottom: bottom, right: right)
  }
}
